pkg/mobile/types.go

```diff
@@ -76,23 +95,46 @@ func (m *Metric) Validate() (valid bool, reason string) {
 		return false, missingDataError
 	}
 
-	if m.Data.Security != nil {
-		if len(*m.Data.Security) == 0 {
-			return false, securityMetricsEmptyError
+	switch m.EventType {
+	case "init":
+		return validateInitMetric(m.Data)
```

Comment:
This approach assumes that the security event type would not contain App or Device metrics, whereas they definitely will very soon.

As the number of event types grows, I can see this approach leading to headaches and lots of code repetition to try fix those headaches.

The Metrics endpoint allows any and all fields defined in the Metric strcut. Instead of explicitly checking the event type and then conditionally checking fields related to that type, we should validate any field if it is present.

Let's go back to the original validate function. If a security field is present, validate it. If an App field is present, validate it. If a Device field is present, validate it.

The only change that should have been required here is 

```go
if m.EventTpe == "" {
	return false, missingEventTypeError
}
```

What do you think @paolobueno?

Reply:
Let's keep things simple for now. In the future I can see each metric struct having its own Validate() function

i.e., the AppMetric will have its own Validate() function, So will DeviceMetric and SecurityMetric

and then in the top level Validate() function belonged to the Metric struct we will do something like this:

```
if m.App {
	if valid, reason := m.App.Validate(); !valid {
}
```

Reply:
Let's keep things simple for now. In the future I can see each metric struct having its own Validate() function

i.e., the AppMetric will have its own Validate() function, So will DeviceMetric and SecurityMetric

and then in the top level Validate() function belonged to the Metric struct we will do something like this:

```

func (m *Metric) Validate(boolean, string) {

	// validate stuff like clientId and eventType

	if m.App != nil {
		if valid, reason := m.App.Validate(); !valid {
			return valid, reason
		}
	}

	if m.Device != nil {
		if valid, reason := m.Device.Validate(); !valid {
			return valid, reason
		}
	}

	if m.Security != nil {
		if valid, reason := m.Security.Validate(); !valid {
			return valid, reason
		}
	}
}
```

I think this approach is the next iteration we should take on validation. If we go down the path of performing different sets of validations based on eventType we are going to be very sad in a few months 😄

Reply:
:+1:  Going ahead with implementing this shape for the data validation because it's unrelated to how we decide to handle the `type` field